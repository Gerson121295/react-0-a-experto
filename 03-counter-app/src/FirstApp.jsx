import PropTypes from 'prop-types'


//Recomendable agregar variables o funciones que no tengan nada que ver con la funcionalidad a fuera de la funcion 
//const newMessage = 'Gerson';
//const newMessage = [1,2,3,4,5,6,7];

/* const newMessage = {
     message: 'Hola',
    title: 'Gep'
} 
*/

function suma(a,b){
    return a + b;
}

export const FirstApp = ({title='No hay title', subTitle='No hay subtitle', name='Gema'}) =>{ //recibe props del padre main.jsx: {title} esta desestructurado de props

    return(
        <>
            {/* <h1>GErson Ep</h1> */}
            {/* <p>{newMessage.message}</p> */} {/* impresion de variables en el HTML */}
            {/* <code>{JSON.stringify(newMessage)}</code>  */}{/* impresion de objeto js */}
            {/* {<p>La suma es: {suma(5,5)}</p>} */}

            {/* Recibe datos mediante props */}
            {/* Title es recibida como props */}
            <h1 data-testid="test-title">{title}</h1> 
            <p>{subTitle}</p> {/* <p>{subTitle} {suma(5,5)}</p>  */}
            <p>{subTitle}</p> 
            <h2>{name}</h2>
        </>
    )
}

/*
export function FirstComponent(){
    return(
        <h3>First component number 2</h3>
    )
}
*/

//Validar las props que se recibe del padre cumpla con el tipo de dato requerido
FirstApp.propTypes ={
    //name: PropTypes.string,
    title: PropTypes.string.isRequired,
    subTitle: PropTypes.string,
    name: PropTypes.string,
    
}


//Definir los props por defecto en caso de que no viene del padre pero si viene tomará la del padre
//Para futuras versiones ya no se utilizará .defaultProps = {  
//para establecer props por defecto se realizará en la funcion donde se define la props a recibir del padre:
//export const FirstApp = ({ title = 'No hay título', subTitle = 'No hay subtítulo' }) => {...}
/*
FirstApp.defaultProps = {
    name: 'Gerson Ep',
    subTitle: 'No hay subtitulo',
    //title: "No hay titulo"
}
*/
export default FirstApp;