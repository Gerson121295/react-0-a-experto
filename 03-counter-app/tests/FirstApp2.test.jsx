import { render, screen } from "@testing-library/react"; //screen ayuda a hacer limpiezas automaticas depues de pruebas
import FirstApp from "../src/FirstApp";

describe('Prueba en <FirstApp /> ', () => {

    //Variables globales para todas las test
    const title = 'Hola, Soy goku';
    const subTitle = 'Hola, Soy un subtitle';

    test('Debe hacer match con el snapshot', () => {

        const {container} = render(
            <FirstApp 
                title={title}
            />
        );

        expect(container).toMatchSnapshot(); //crea el snapshot
        
    });

    test('debe de mostrar el mensaje "Hola Soy Goku', () => {

        render(
            <FirstApp 
                title={title}
            />
        );

        //expect(screen.getByText(title)).not.toBeTruthy();  //valida que no exista
        expect(screen.getByText(title)).toBeTruthy();  //que exista
        

    });

    test('debe de mostrar el titulo en un h1', () => {
        render(
            <FirstApp 
                title={title}
            />
        );

        //muestra el title en h1 heading level 1
        expect(screen.getByRole('heading', {level: 1}).innerHTML ).toContain(title);
    });

    test('debe de mostrar el subtitulo enviado por props', () => {
        render(
            <FirstApp 
                title={title}
                subTitle={subTitle}
            />
        );
        expect(screen.getAllByText(subTitle).length).toBe(2);

    });
});

