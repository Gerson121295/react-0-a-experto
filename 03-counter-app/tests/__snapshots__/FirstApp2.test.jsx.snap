// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <FirstApp />  Debe hacer match con el snapshot 1`] = `
<div>
   
  <h1
    data-testid="test-title"
  >
    Hola, Soy goku
  </h1>
  <p>
    No hay subtitle
  </p>
   
  <p>
    No hay subtitle
  </p>
  <h2>
    Gema
  </h2>
</div>
`;
